packages/flutter/lib/src/gestures/pointer_data_resampler.dart

```diff
@@ -0,0 +1,199 @@
+  final _queuedData = <ui.PointerData>[];
+
+  /// Pointer state required for resampling.
+  ui.PointerData _last;
+  ui.PointerData _next;
+  var _position = ui.Offset(0.0, 0.0);
+  var _isTracked = false;
+  var _isDown = false;
+  var _pointerIdentifier = 0;
+
+  ui.PointerData _clone(
+      ui.PointerData data,
+      ui.PointerChange change,
+      ui.Offset position,
+      ui.Offset delta,
+      int pointerIdentifier,
+      Duration timeStamp) {
+    return ui.PointerData(
+        buttons: data.buttons,
+        device: data.device,
+        timeStamp: timeStamp,
+        change: change,
+        kind: data.kind,
+        physicalDeltaX: delta.dx,
+        physicalDeltaY: delta.dy,
+        physicalX: position.dx,
+        physicalY: position.dy,
+        pointerIdentifier: pointerIdentifier,
+        synthesized: data.synthesized);
+  }
+
+  ui.Offset _positionAt(Duration sampleTime) {
+    // Use `next` position by default.
+    var x = _next.physicalX;
+    var y = _next.physicalY;
+
+    // Resample if `next` time stamp is past `sampleTime`.
+    if (_next.timeStamp > sampleTime && _next.timeStamp > _last.timeStamp) {
+      final interval =
+          (_next.timeStamp - _last.timeStamp).inMicroseconds.toDouble();
+      final scalar =
+          (sampleTime - _last.timeStamp).inMicroseconds.toDouble() / interval;
+      x = _last.physicalX + (_next.physicalX - _last.physicalX) * scalar;
+      y = _last.physicalY + (_next.physicalY - _last.physicalY) * scalar;
+    }
+
+    return ui.Offset(x, y);
+  }
+
+  void _processPointerData(Duration sampleTime) {
+    for (var i = 0; i < _queuedData.length; i++) {
+      final data = _queuedData[i];
+
+      // Update both `last` and `next` pointer data if time stamp is older
+      // or equal to `sampleTime`.
+      if (data.timeStamp <= sampleTime || _last == null) {
+        _last = data;
+        _next = data;
+        continue;
+      }
+
+      // Update only `next` pointer data if time stamp is more recent than
+      // `sampleTime` and next data is not already more recent.
+      if (_next.timeStamp < sampleTime) {
+        _next = data;
+      }
+    }
+  }
+
+  List<ui.PointerData> _dequeueAndSamplePointerDataUntil(Duration sampleTime) {
+    final dequeuedData = <ui.PointerData>[];
+
+    while (_queuedData.isNotEmpty) {
+      final data = _queuedData.first;
+
+      // Potentially stop dequeuing data if more recent than `sampleTime`.
+      if (data.timeStamp > sampleTime) {
+        // Stop if change is not `up` or `remove`, which are allowed to be
+        // processed early as this improves resampling of these changes,
+        // which is important for fling animations.
+        if (data.change != ui.PointerChange.up &&
+            data.change != ui.PointerChange.remove) {
+          break;
+        }
+
+        // Time stamp must match `next` data for early processing to be
+        // allowed.
+        if (data.timeStamp != _next.timeStamp) {
```

Comment:
These expressions are not the same. This expression prevents us from processing data past the _next data. We allow _next data to be processed early for up/remove changes but we need to stop processing if we've gone past _next.

Reply:
Ah yes, I forgot those events past `_next`. I was only thinking the first data with `data.timeStamp > sampleTime`. I think it's probably clearer to write `if (data.timeStamp > _next.timeStamp)` as `_next` is the `data` with the smallest `timeStamp > sampleTime` so we're not expecting any `data` with `sampleTime < data.timeStamp < _next.timeStamp`. This time I tried it with unit tests below to make sure that it's not trivially false like the `if (sampleTime == _next.timeStamp)` condition...

Overall, I wrote the following to dump all my thoughts. From the comments below, it seems that the fling smoothing ("up"/"remove" early processing) would only be guaranteed when the input frequency is no greater than the display frequency?
```
        // When this line is reached, the following two invariants hold:
        // (1) data.timeStamp > sampleTime
        // (2) _next has the smallest timeStamp that's no less than sampleTime
        //
        // Therefore, data must satisfy data.timeStamp >= _next.timeStamp.
        //
        // Those data with the minimum data.timeStamp == _next.timeStamp
        // time stamp are processed early for smoother fling. For data with
        // data.timeStamp > _next.timeStamp, the following lines break the while
        // loop to stop the early processing.
        //
        // Specifically, when sampleTime < _next.timeStamp, there must be at
        // least one data with _next.timeStamp == data.timeStamp > sampleTime
        // and that data is _next itself, and it will be processed early.
        //
        // When sampleTime == _next.timeStamp, all data with data.timeStamp >
        // sampleTime must also have data.timeStamp > _next.timeStamp so no data
        // will be processed early.
        //
        // When the input frequency is no greater than the display
        // frequency, this early processing should guarantee that "up" and
        // "remove" events are always re-sampled (by linear interpolation).
        assert(data.timeStamp >= _next.timeStamp);
        if (data.timeStamp > _next.timeStamp) {
          break;
        }
```

Reply:
Thanks for taking a detailed look at this. I changed to `if (data.timeStamp > _next.timeStamp)` and included your comment (with some minor adjustments) in the code.

Yes, this early processing only works when input frequency is lower than the sampling frequency, which is when this has been determined to be critical. Good call to explain that in a comment.

We could allow the "next" sample time to be passed to the resampler and process up/remove until that time to enable the same for high frequency input events. Some experiments on devices with high frequency input events would be useful to determine the value of this. I can imagine it being useful when input frequency is 2x the refresh rate.